src/common/sys_lmdb.c

```diff
@@ -0,0 +1,511 @@
+
+	unlink(ldb->db_file); /* ignore error code */
+}
+
+static int
+mdb_error2daos_error(int rc)
+{
+	if (rc > 0)
+		rc = -rc;
+
+	switch (rc) {
+	case 0:
+		return 0;
+	case MDB_VERSION_MISMATCH:
+		return -DER_MISMATCH;
+	case MDB_INVALID:
+		return -DER_INVAL;
+	case MDB_PANIC:
+	case MDB_MAP_RESIZED:
+		return -DER_SHUTDOWN;
+	case MDB_READERS_FULL:
+		return -DER_AGAIN;
+	case MDB_NOTFOUND:
+		return -DER_NONEXIST;
+	case MDB_KEYEXIST:
+		return -DER_EXIST;
+	default:
+		return daos_errno2der(-rc);
+	}
+
+}
+
+/* open or create system DB stored in external storage */
+static int
+lmm_db_open_create(struct sys_db *db, bool try_create)
+{
+	struct lmm_sys_db *ldb = db2lmm(db);
+	d_iov_t		   key;
+	d_iov_t		   val;
+	uint32_t	   ver;
+	int		   rc = 0;
+
+	if (try_create) {
+		rc = mkdir(ldb->db_path, 0777);
+		if (rc < 0 && errno != EEXIST) {
+			rc = daos_errno2der(errno);
+			return rc;
+		}
+	} else if (access(ldb->db_file, F_OK) != 0) {
+		D_DEBUG(DB_IO, "%s doesn't exist, bypassing lmdb open\n",
+			ldb->db_file);
+		rc = -DER_NONEXIST;
+		return rc;
+	} else if (access(ldb->db_file, R_OK | W_OK) != 0) {
+		rc = -DER_NO_PERM;
+		D_CRIT("No access to existing db file %s\n", ldb->db_file);
+		return rc;
+	}
+
+	D_DEBUG(DB_IO, "Opening %s, try_create=%d\n", ldb->db_file, try_create);
+	rc = mdb_env_create(&ldb->db_env);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to create env handle for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_env_open(ldb->db_env, ldb->db_file, MDB_NOSUBDIR, 0664);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to open env handle for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_txn_begin(ldb->db_env, NULL, 0, &ldb->db_txn);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to begin tx for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_dbi_open(ldb->db_txn, NULL, 0, &ldb->db_dbi);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to open sysdb: "DF_RC"\n", DP_RC(rc));
+		goto txn_abort;
+	}
+
+
+	d_iov_set(&key, SYS_DB_MD_VER, strlen(SYS_DB_MD_VER));
+	d_iov_set(&val, &ver, sizeof(ver));
+	if (try_create) {
+		ver = SYS_DB_VERSION;
+		rc = lmm_db_upsert(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to set version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+			goto txn_abort;
+		}
+		rc = mdb_txn_commit(ldb->db_txn);
+		if (rc)
+			D_CRIT("Failed to commit version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+		goto out;
+	} else {
+		rc = lmm_db_fetch(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to read sysdb version: "DF_RC"\n",
+			       DP_RC(rc));
+			rc = -DER_INVAL;
+		}
+
+		if (ver < SYS_DB_VERSION_1 || ver > SYS_DB_VERSION)
+			rc = -DER_DF_INCOMPT;
+	}
+txn_abort:
+	mdb_txn_abort(ldb->db_txn);
+out:
+	ldb->db_txn = NULL;
+	return rc;
+}
+
+#define MAX_SMD_TABLE_LEN	32
+static int
+lmm_db_generate_key(char *table, d_iov_t *key, MDB_val *db_key)
```

Comment:
Instead of generating new keys in this way, what about using multiple dbs? (MDB_dbi per table, can single lmdb file support multiple MDB_dbi?)

Reply:
I tried, it almost work, but when i killed daos server and restart, data lost. i see docs:

"Database names are keys in the unnamed database, and may be read but not written."

then i reverted to this way, it could work anyway.